### PR TITLE
Finetune for eval tasks

### PR DESCRIPTION
**Pull request by yawenzzzz**
Support finetuning Helios & other models on our eval tasks. [Results](https://wandb.ai/eai-ai2/2025_09_18_phase1_finetune?nw=nwuseryawenz). 
Example: 
```
python /weka/dfive-default/yawenz/helios/helios/internal/full_eval_sweep.py --checkpoint_path /weka/dfive-default/helios/checkpoints/yawenzzzz/base_v6_default/step400000 --project_name 2025_09_18_phase1_finetune --module_path /weka/dfive-default/yawenz/helios/scripts/2025_09_10_phase1/script.py --cluster ai2/ceres --finetune
```
Beaker job: https://beaker.allen.ai/orgs/ai2/workspaces/earth-systems/work/01K5W6R2G0HRPNE7GCZZENYYQD?taskId=01K5W6R2GXMG8AQP5K0WQJJAY1&jobId=01K5W6R2MPM3ASSGJKF2EH4SQM

I created `FT_EVAL_TASKS`, a subset of our evaluation tasks covering classification, segmentation, and single-pixel classification. I’m not sure if we should run FT on all tasks, but like KNN and LP, it would be useful to see how performance changes with sample size. For tasks like PASTIS, FT is already supported by rslp. This launcher is mainly for tasks outside rslearn projects, so we don’t need to convert datasets into rslearn format to run finetuning.

### CODE REVIEW

File: helios/evals/finetune.py [hunk 1]
```diff
@@ -1,742 +1,239 @@
...
+        use_pooled_tokens=use_pooled_tokens,
     ).to(device)
 
-    finetuned_model = finetuned_model.train()
-    optimizer = torch.optim.AdamW(finetuned_model.parameters(), lr=lr)
-
-    loss_function = nn.CrossEntropyLoss(ignore_index=-1)
+    # Trigger _init_head once with a tiny dry pass
+    with torch.no_grad(), torch.autocast(device_type=device.type, dtype=torch.bfloat16):
+        sample_batch, _ = next(iter(train_loader))
```
Comment (Hgherzog) [inline helios/evals/finetune.py]: This skips the first sample in the dataset. It would be preferable to just get the information about how to set up the head based on the task config, whether it is segmentation and the model output dimension

File: helios/evals/finetune.py [hunk 2]
```diff
@@ -1,742 +1,239 @@
...
-            self.spatial_pool = True
+        self.num_classes = num_classes
+        # placeholder head; real in_dim discovered on first forward
+        self._head = nn.Linear(1, 1, bias=True)
+        self._inited = False
+
+    def _init_head(self, emb_dim: int, device: torch.device) -> None:
+        """Initialize the head based on the embedding dimension."""
+        if self.task_type == TaskType.CLASSIFICATION:
+            self._head = nn.Linear(emb_dim, self.num_classes, bias=True)
```
Comment (Hgherzog) [inline helios/evals/finetune.py:56]: We should add a batch-norm layer before the linear head, as we do for the linear probing for classification. This was shown to make evaluations more robust in this paper https://arxiv.org/html/2407.12210v2

File: helios/internal/all_evals.py [hunk 3]
```diff
@@ -172,7 +184,7 @@ def load_user_module(path: str) -> Any:
     ),
     "pastis_sentinel1_sentinel2": DownstreamTaskConfig(
         dataset="pastis",
-        embedding_batch_size=32,
+        embedding_batch_size=512,
```
Comment (Hgherzog) [inline helios/internal/all_evals.py:156]: Revert this.

File: helios/internal/all_evals.py [hunk 4]
```diff
@@ -297,6 +309,62 @@ def load_user_module(path: str) -> Any:
     ),
 }
 
+# TODO: We can add other classification tasks once they are cut into 128*128
```
Comment (Hgherzog) [inline helios/internal/all_evals.py]: Why can't we include big earthnet and so2-sat here? I would also be curious how pastis does with just a linear layer

File: helios/internal/full_eval_sweep.py [hunk 5]
```diff
@@ -13,28 +13,52 @@
...
-from helios.internal.all_evals import EVAL_TASKS
+from helios.internal.all_evals import EVAL_TASKS, FT_EVAL_TASKS
 from helios.internal.experiment import SubCmd
 from helios.nn.flexihelios import PoolingType
 
+# Linear probe learning rates to sweep over
 LP_LRs = [1e-4, 5e-4, 1e-3, 5e-3, 1e-2, 5e-2, 1e-1, 5e-1]
-Normalization_MODES = ["dataset", "pre_trained"]
+# Fine-tune learning rates to sweep over
+FT_LRs = [1e-5, 3e-5, 6e-5, 1e-4, 3e-4, 6e-4, 1e-3]
```
Comment (Hgherzog) [inline helios/internal/full_eval_sweep.py]: I would consider extending this range a bit higher.

File: helios/internal/full_eval_sweep.py [hunk 6]
```diff
@@ -13,28 +13,52 @@
...
+
+Normalization_MODES = ["pre_trained", "dataset"]
```
Comment (Hgherzog) [inline helios/internal/full_eval_sweep.py]: Nit: Make this a StrEnum

File: helios/internal/full_eval_sweep.py [hunk 7]
```diff
@@ -87,67 +111,116 @@ def no_norm_sweep() -> Generator[dict[str, Any], None, None]:
...
     # Normalization is then applied by the eval wrapper by default
-    dino_v3_args = dataset_args
-    dino_v3_args += " " + " ".join(
-        [
-            f"--trainer.callbacks.downstream_evaluator.tasks.{task_name}.norm_method=NormMethod.NORM_YES_CLIP_MIN_MAX_INT"
-            for task_name in EVAL_TASKS.keys()
-        ]
-    )
+    dino_v3_args = dataset_args if not finetune else ft_dataset_args
+    if not finetune:
```
Comment (Hgherzog) [inline helios/internal/full_eval_sweep.py]: Suggestion: Condense the fine_tuning logic to the following code for all the different models
`tasks = FT_EVAL_TASKS if finetune else EVAL_TASKS 
dino_v3_args += " " + " ".join(
            [
                f"--trainer.callbacks.downstream_evaluator.tasks.{task_name}.norm_method=NormMethod.NORM_YES_CLIP_MIN_MAX_INT"
                for task_name in tasks.keys()
            ]
        )`

File: helios/train/callbacks/evaluator_callback.py [hunk 8]
```diff
@@ -100,10 +107,10 @@ def __init__(
         if self.eval_mode is None:
             self.eval_mode = get_eval_mode(self.config.task_type)
 
-        assert self.eval_mode in [
-            "knn",
-            "linear_probe",
-        ], f"Unexpected eval mode {self.eval_mode}"
+        assert self.eval_mode in ["knn", "linear_probe", "finetune"], (
```
Comment (Hgherzog) [inline helios/train/callbacks/evaluator_callback.py]: Nit: Make these mdoes into StrEnum

File: helios/train/masking.py
Comment (Hgherzog) [inline helios/train/masking.py:1]: These changes should be merged first in the phase 1 PR

File: helios/internal/all_evals.py (hunk 3 above)
Comment (yawenzzzz) [reply helios/internal/all_evals.py:156]: Fixed.

File: helios/internal/all_evals.py (hunk 4 above)
Comment (yawenzzzz) [reply helios/internal/all_evals.py]: I got OOM with Helios model as mentioned before, and will try not loading the optim state to see if this resolve the issue.

File: helios/internal/full_eval_sweep.py (hunk 5 above)
Comment (yawenzzzz) [reply helios/internal/full_eval_sweep.py]: OK, extended to 1e-2. For finetuning, a lower LR is usually better, but we can test and see.

File: helios/internal/full_eval_sweep.py (hunk 6 above)
Comment (yawenzzzz) [reply helios/internal/full_eval_sweep.py]: Fixed.

File: helios/train/callbacks/evaluator_callback.py (hunk 8 above)
Comment (yawenzzzz) [reply helios/train/callbacks/evaluator_callback.py]: Fixed.

Comment (yawenzzzz) [reply helios/train/callbacks/evaluator_callback.py]: Fixed.

File: helios/internal/full_eval_sweep.py (hunk 7 above)
Comment (yawenzzzz) [reply helios/internal/full_eval_sweep.py]: Fixed, thanks!

File: helios/evals/finetune.py (hunk 2 above)
Comment (yawenzzzz) [reply helios/evals/finetune.py:56]: Added the BatchNorm layer, in Linear Probing, we use BatchNorm only in classification task not in segmentation task.

File: helios/evals/finetune.py (hunk 1 above)
Comment (yawenzzzz) [reply helios/evals/finetune.py]: This is not the real training loop, the goal here is to use a tiny dry run to initialize the head based on the ft model that was initialized. Given that task type is input of ft, the model is already set up and this is mainly for trigger `_init_head`.

File: helios/internal/all_evals.py (hunk 4 above)
Comment (yawenzzzz) [reply helios/internal/all_evals.py]: Will add them later.

File: helios/evals/datasets/configs.py [hunk 9]
```diff
@@ -4,6 +4,7 @@
 from typing import NamedTuple
 
 from helios.data.constants import Modality
+from helios.train.callbacks.evaluator_callback import EvalMode
```
[the remark this answers is not in the document]
Comment (yawenzzzz) [reply helios/evals/datasets/configs.py]: Fixed.

File: helios/evals/finetune.py [hunk 10]
```diff
@@ -1,742 +1,318 @@
...
-
-    return finetuned_model
+    def forward(self, batch: MaskedHeliosSample) -> torch.Tensor:
+        """Forward pass through the model and head."""
+        dev = next(self.wrapper.parameters()).device
+        # classification: (B, D), segmentation: (B, H, W, D)
+        emb, _ = self.wrapper(batch, None)
+        emb = cast(torch.Tensor, emb)
+        emb_dim = emb.shape[-1]
+        self.batch_norm = nn.BatchNorm1d(emb_dim).to(device=dev)
```
[the remark this answers is not in the document]
Comment (yawenzzzz) [reply helios/evals/finetune.py]: Fixed.

File: helios/train/callbacks/evaluator_callback.py [hunk 11]
```diff
@@ -392,7 +485,7 @@ class DownstreamEvaluatorCallbackConfig(CallbackConfig):
     cancel_after_first_eval: bool = False
     tasks_to_run: list[str] | None = None
     # whether to run the evaluators on the val set only (=False) or on the test and val set (=True)
-    run_on_test: bool = False
+    run_on_test: bool = True
```
[the remark this answers is not in the document]
Comment (yawenzzzz) [reply helios/train/callbacks/evaluator_callback.py]: Fixed.
